src/Elastic.Apm/Metrics/MetricsProvider/CgroupMetricsProvider.cs

```diff
@@ -0,0 +1,344 @@
+		private const string Cgroup1Unlimited = "9223372036854771712";
+		private const string Cgroup2Unlimited = "max";
+
+		internal const string SystemProcessCgroupMemoryMemLimitBytes = "system.process.cgroup.memory.mem.limit.bytes";
+		internal const string SystemProcessCgroupMemoryMemUsageBytes = "system.process.cgroup.memory.mem.usage.bytes";
+		internal const string SystemProcessCgroupMemoryStatsInactiveFileBytes = "system.process.cgroup.memory.stats.inactive_file.bytes";
+
+		internal static readonly Regex MemoryCgroup = new Regex("^\\d+:memory:.*");
+		internal static readonly Regex Cgroup1MountPoint = new Regex("^\\d+? \\d+? .+? .+? (.*?) .*cgroup.*memory.*");
+		internal static readonly Regex Cgroup2MountPoint = new Regex("^\\d+? \\d+? .+? .+? (.*?) .*cgroup2.*cgroup.*");
+
+		private readonly CgroupFiles _cGroupFiles;
+
+		/// <summary>
+		/// Initializes a new instance of <see cref="CgroupMetricsProvider"/>
+		/// </summary>
+		/// <param name="logger">the logger</param>
+		/// <param name="collectMemLimitBytes">whether to collect <see cref="SystemProcessCgroupMemoryMemLimitBytes"/> metric</param>
+		/// <param name="collectMemUsageBytes">whether to collect <see cref="SystemProcessCgroupMemoryMemUsageBytes"/> metric</param>
+		/// <param name="collectStatsInactiveFileBytes">whether to collect <see cref="SystemProcessCgroupMemoryStatsInactiveFileBytes"/> metric</param>
+		public CgroupMetricsProvider(IApmLogger logger, bool collectMemLimitBytes = true, bool collectMemUsageBytes = true, bool collectStatsInactiveFileBytes = true)
+			: this(ProcSelfCgroup, ProcSelfMountinfo, logger, collectMemLimitBytes, collectMemUsageBytes, collectStatsInactiveFileBytes)
+		{
+		}
+
+		/// <summary>
+		/// Initializes a new instance of <see cref="CgroupMetricsProvider"/>
+		/// </summary>
+		/// <param name="procSelfCGroup">the <see cref="ProcSelfCgroup"/> file</param>
+		/// <param name="mountInfo">the <see cref="ProcSelfMountinfo"/> file</param>
+		/// <param name="logger">the logger</param>
+		/// <param name="collectMemLimitBytes">whether to collect <see cref="SystemProcessCgroupMemoryMemLimitBytes"/> metric</param>
+		/// <param name="collectMemUsageBytes">whether to collect <see cref="SystemProcessCgroupMemoryMemUsageBytes"/> metric</param>
+		/// <param name="collectStatsInactiveFileBytes">whether to collect <see cref="SystemProcessCgroupMemoryStatsInactiveFileBytes"/> metric</param>
+		/// <remarks>
+		///	Used for testing
+		/// </remarks>
+		internal CgroupMetricsProvider(string procSelfCGroup, string mountInfo, IApmLogger logger, bool collectMemLimitBytes = true, bool collectMemUsageBytes = true, bool collectStatsInactiveFileBytes = true)
+		{
+			_collectMemLimitBytes = collectMemLimitBytes;
+			_collectMemUsageBytes = collectMemUsageBytes;
+			_collectStatsInactiveFileBytes = collectStatsInactiveFileBytes;
+			_logger = logger.Scoped(nameof(CgroupMetricsProvider));
+			_cGroupFiles = FindCGroupFiles(procSelfCGroup, mountInfo);

```

Comment:
I was first thinking that maybe I'd try catch this call. Reason I'm super careful here is that this is directly called from `MetricsCollector` `.ctor` so if an exception is thrown here, it would crash the application. But then in `FindCGroupFiles` all file reading part is try catched, so I think we are good here.